content/en/blog/_posts/2023-09-15-bootstrap-an-air-gapped-cluster-with-kubeadm/index.md

```diff
@@ -0,0 +1,681 @@
+---
+layout: blog
+title: "Bootstrap an Air Gapped Cluster With Kubeadm"
+date: 2023-09-15
+slug: bootstrap-an-air-gapped-cluster-with-kubeadm
+---
+
+**Author:** Rob Mengert (Defense Unicorns)
+
+Ever wonder how software gets deployed onto a system that is deliberately disconnected from the Internet and other networks? These systems are typically disconnected due to their sensitive nature. Sensitive as in utilities (power/water), banking, healthcare, weapons systems, other government use cases, etc. Sometimes it's technically a water gap, if you're running Kubernetes on an underwater vessel. Still, these environments need software to operate. This concept of deployment in a disconnected state is what it means to deploy to the other side of an [air gap](https://en.wikipedia.org/wiki/Air_gap_(networking)).
+
+Again, despite this posture, software still needs to run in these environments. Traditionally, software artifacts are physically carried across the air gap on hard drives, USB sticks, CDs, or floppy disks (for ancient systems, it still happens). Kubernetes lends itself particularly well to running software behind an air gap for several reasons, largely due to its declarative nature.
+
+In this blog article, I will walk through the process of bootstrapping a Kubernetes
+cluster in an air-gapped lab environment using Fedora Linux and kubeadm.
+
+## The Air Gap VM Setup
+
+A real air-gapped network can take some effort to set up, so for this post, I will use an example VM on a laptop and do some network modifications. Below is the topology:
+
+{{< figure src="airgap-vm.svg" alt="Topology on the host/laptop which shows that connectivity to the internet from the air gap VM is not possible. However, connectivity between the host/laptop and the VM is possible" >}}
+
+### Local Topology
+
+This VM will have its network connectivity disabled but in a way that doesn't shut down the VM's virtual NIC. Instead, its network will be downed by injecting a default route to a dummy interface, making anything internet-hosted unreachable. However, the VM still has a connected route to the bridge interface on the host, which means that network connectivity to the host is still working. This posture means that data can be transferred from the host/laptop to the VM via scp, even with the default route on the VM black-holing all traffic that isn't destined for the local bridge subnet. This type of transfer is analogous to carrying data across the air gap and will be used throughout this post.
+
+Other details about the lab setup:
+
+**VM OS:** Fedora 37
+
+**Kubernetes Version:** v1.27.3
+
+**CNI Plugins Version:** v1.3.0
+
+**CNI Provider and Version:** Flannel v0.22.0
```

Comment:
(nit)
```suggestion
**VM OS:** Fedora 37  
**Kubernetes Version:** v1.27.3  
**CNI Plugins Version:** v1.3.0  
**CNI Provider and Version:** Flannel v0.22.0  
```

There are `  ` two spaces at the end of each line here. Our Markdown renderer turns those into hard breaks.